File: services/identity-service/src/test/java/br/com/deltaglobalbank/identity/features/apiClient/createApiClient/CreateApiClientTests.java
package br.com.deltaglobalbank.identity.features.apiClient.createApiClient;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.util.List;
import java.util.UUID;

import br.com.deltaglobalbank.identity.domain.apiClient.ApiClient;
import br.com.deltaglobalbank.identity.domain.apiClient.ApiClientRepository;
import br.com.deltaglobalbank.identity.domain.module.ModuleRepository;
import br.com.deltaglobalbank.identity.domain.module.TenantModule;
import br.com.deltaglobalbank.identity.domain.module.TenantModuleRepository;
import br.com.deltaglobalbank.identity.domain.role.Role;
import br.com.deltaglobalbank.identity.domain.role.RoleCode;
import br.com.deltaglobalbank.identity.domain.role.RoleInactiveException;
import br.com.deltaglobalbank.identity.domain.role.RoleNotFoundException;
import br.com.deltaglobalbank.identity.domain.role.RoleRepository;
import br.com.deltaglobalbank.identity.domain.tenant.Tenant;
import br.com.deltaglobalbank.identity.domain.tenant.TenantRepository;
import br.com.deltaglobalbank.identity.domain.user.DuplicateRoleException;
import br.com.deltaglobalbank.identity.domain.user.ModuleNotEnabledForTenantException;
import br.com.deltaglobalbank.identity.domain.user.TenantInactiveException;
import br.com.deltaglobalbank.identity.domain.user.TenantNotFoundException;
import br.com.deltaglobalbank.identity.features.apiClients.createApiClient.CreateApiClientCommand;
import br.com.deltaglobalbank.identity.features.apiClients.createApiClient.CreateApiClientResponse;
import br.com.deltaglobalbank.identity.features.apiClients.createApiClient.CreateApiClientUseCase;
import br.com.deltaglobalbank.identity.infrastructure.persistence.repositories.JpaApiClientRoleRepository;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.security.authorization.AuthorizationDeniedException;

class CreateApiClientTests {

    private final ApiClientRepository apiClientRepository = mock(ApiClientRepository.class);
    private final TenantRepository tenantRepository = mock(TenantRepository.class);
    private final RoleRepository roleRepository = mock(RoleRepository.class);
    private final TenantModuleRepository tenantModuleRepository = mock(TenantModuleRepository.class);
    private final ModuleRepository moduleRepository = mock(ModuleRepository.class);
    private final JpaApiClientRoleRepository apiClientRoleRepository = mock(JpaApiClientRoleRepository.class);

    private CreateApiClientUseCase useCase;

    @BeforeEach
    void setUp() {
        useCase = new CreateApiClientUseCase(
            apiClientRepository, tenantRepository, roleRepository,
            tenantModuleRepository, moduleRepository, apiClientRoleRepository
        );
    }

    @Test
    void mustCreateApiClientSuccessfullyWithRoles() {
        UUID tenantId = UUID.randomUUID();
        RoleCode roleCode = new RoleCode("customers.admin");
        CreateApiClientCommand command = new CreateApiClientCommand(
            tenantId, "ERP-client", "ERP", List.of(roleCode), List.of("platform.admin"));

        Tenant tenant = mock(Tenant.class);
        when(tenant.isActive()).thenReturn(true);
        when(tenantRepository.findById(tenantId)).thenReturn(tenant);

        UUID moduleId = UUID.randomUUID();
        Role role = mock(Role.class);
        when(role.getCode()).thenReturn(roleCode);
        when(role.getModuleId()).thenReturn(moduleId);
        when(role.getId()).thenReturn(UUID.randomUUID());
        when(role.isActive()).thenReturn(true);
        when(roleRepository.findByCode(roleCode)).thenReturn(role);

        TenantModule tenantModule = mock(TenantModule.class);
        when(tenantModule.getModuleId()).thenReturn(moduleId);
        when(tenantModuleRepository.findAllByTenantIdAndEnabled(tenantId, true)).thenReturn(List.of(tenantModule));

        when(apiClientRepository.save(any())).thenReturn(mock(ApiClient.class));

        CreateApiClientResponse response = useCase.execute(command);

        Assertions.assertAll(
            () -> assertEquals(tenantId, response.tenantId()),
            () -> assertEquals("ERP-client", response.name()),
            () -> assertEquals("ERP", response.description())
        );

        verify(apiClientRepository, times(1)).save(any());
        verify(apiClientRoleRepository, times(1)).save(any());
    }

    @Test
    void mustThrowTenantNotFoundExceptionWhenTenantDoesNotExist() {
        CreateApiClientCommand command = new CreateApiClientCommand(
            UUID.randomUUID(), "ERP-client", null, List.of(new RoleCode("customers.admin")), List.of("platform.admin"));

        when(tenantRepository.findById(any())).thenReturn(null);

        assertThrows(TenantNotFoundException.class, () -> useCase.execute(command));

        verify(apiClientRepository, times(0)).save(any());
        verify(apiClientRoleRepository, times(0)).save(any());
    }

    @Test
    void mustThrowTenantInactiveExceptionWhenTenantIsInactive() {
        UUID tenantId = UUID.randomUUID();
        CreateApiClientCommand command = new CreateApiClientCommand(
            tenantId, "ERP-client", null, List.of(new RoleCode("customers.admin")), List.of("platform.admin"));

        Tenant tenant = mock(Tenant.class);
        when(tenant.isActive()).thenReturn(false);
        when(tenantRepository.findById(tenantId)).thenReturn(tenant);

        assertThrows(TenantInactiveException.class, () -> useCase.execute(command));

        verify(apiClientRepository, times(0)).save(any());
        verify(apiClientRoleRepository, times(0)).save(any());
    }

    @Test
    void mustThrowDuplicateRoleExceptionWhenRoleCodesAreExactlyDuplicated() {
        UUID tenantId = UUID.randomUUID();
        RoleCode roleCode = new RoleCode("customers.admin");
        CreateApiClientCommand command = new CreateApiClientCommand(
            tenantId, "ERP-client", null, List.of(roleCode, roleCode), List.of("platform.admin"));

        Tenant tenant = mock(Tenant.class);
        when(tenant.isActive()).thenReturn(true);
        when(tenantRepository.findById(tenantId)).thenReturn(tenant);

        assertThrows(DuplicateRoleException.class, () -> useCase.execute(command));

        verify(apiClientRepository, times(0)).save(any());
        verify(apiClientRoleRepository, times(0)).save(any());
    }

    @Test
    void mustThrowDuplicateRoleExceptionWhenRoleCodesDifferOnlyByCase() {
        UUID tenantId = UUID.randomUUID();
        CreateApiClientCommand command = new CreateApiClientCommand(
            tenantId, "ERP-client", null,
            List.of(new RoleCode("CUSTOMERS.ADMIN"), new RoleCode("customers.admin")), List.of("platform.admin"));

        Tenant tenant = mock(Tenant.class);
        when(tenant.isActive()).thenReturn(true);
        when(tenantRepository.findById(tenantId)).thenReturn(tenant);

        assertThrows(DuplicateRoleException.class, () -> useCase.execute(command));

        verify(apiClientRepository, times(0)).save(any());
        verify(apiClientRoleRepository, times(0)).save(any());
    }

    @Test
    void mustThrowRoleNotFoundExceptionWhenRoleDoesNotExist() {
        UUID tenantId = UUID.randomUUID();
        RoleCode roleCode = new RoleCode("nonexistent.role");
        CreateApiClientCommand command = new CreateApiClientCommand(
            tenantId, "ERP-client", null, List.of(roleCode), List.of("platform.admin"));

        Tenant tenant = mock(Tenant.class);
        when(tenant.isActive()).thenReturn(true);
        when(tenantRepository.findById(tenantId)).thenReturn(tenant);

        when(roleRepository.findByCode(roleCode)).thenReturn(null);

        assertThrows(RoleNotFoundException.class, () -> useCase.execute(command));

        verify(apiClientRepository, times(0)).save(any());
        verify(apiClientRoleRepository, times(0)).save(any());
    }

    @Test
    void mustThrowModuleNotEnabledForTenantExceptionWhenRoleModuleIsNotEnabled() {
        UUID tenantId = UUID.randomUUID();
        RoleCode roleCode = new RoleCode("customers.admin");
        CreateApiClientCommand command = new CreateApiClientCommand(
            tenantId, "ERP-client", null, List.of(roleCode), List.of("platform.admin"));

        Tenant tenant = mock(Tenant.class);
        when(tenant.isActive()).thenReturn(true);
        when(tenantRepository.findById(tenantId)).thenReturn(tenant);

        UUID moduleId = UUID.randomUUID();
        Role role = mock(Role.class);
        when(role.getCode()).thenReturn(roleCode);
        when(role.getModuleId()).thenReturn(moduleId);
        when(role.isActive()).thenReturn(true);
        when(roleRepository.findByCode(roleCode)).thenReturn(role);

        when(tenantModuleRepository.findAllByTenantIdAndEnabled(tenantId, true)).thenReturn(List.of());
        when(moduleRepository.findById(moduleId)).thenReturn(null);

        assertThrows(ModuleNotEnabledForTenantException.class, () -> useCase.execute(command));

        verify(apiClientRepository, times(0)).save(any());
        verify(apiClientRoleRepository, times(0)).save(any());
    }

    @Test
    void mustDenyWhenNonPlatformAdminCreatesApiClientWithPlatformAdminRole() {
        UUID tenantId = UUID.randomUUID();
        RoleCode roleCode = new RoleCode("platform.admin");
        CreateApiClientCommand command = new CreateApiClientCommand(
            tenantId, "ERP-client", null, List.of(roleCode), List.of("identity.admin"));

        Tenant tenant = mock(Tenant.class);
        when(tenant.isActive()).thenReturn(true);
        when(tenantRepository.findById(tenantId)).thenReturn(tenant);

        Role role = mock(Role.class);
        when(role.getCode()).thenReturn(roleCode);
        when(role.getModuleId()).thenReturn(null);
        when(role.isActive()).thenReturn(true);
        when(roleRepository.findByCode(roleCode)).thenReturn(role);

        AuthorizationDeniedException ex = assertThrows(AuthorizationDeniedException.class, () -> useCase.execute(command));
        assertEquals("denied", ex.getMessage());

        verify(apiClientRepository, times(0)).save(any());
        verify(apiClientRoleRepository, times(0)).save(any());
    }

    @Test
    void mustAllowPlatformAdminToCreateApiClientWithPlatformAdminRole() {
        UUID tenantId = UUID.randomUUID();
        RoleCode roleCode = new RoleCode("platform.admin");
        CreateApiClientCommand command = new CreateApiClientCommand(
            tenantId, "ERP-client", null, List.of(roleCode), List.of("platform.admin"));

        Tenant tenant = mock(Tenant.class);
        when(tenant.isActive()).thenReturn(true);
        when(tenantRepository.findById(tenantId)).thenReturn(tenant);

        Role role = mock(Role.class);
        when(role.getCode()).thenReturn(roleCode);
        when(role.getModuleId()).thenReturn(null);
        when(role.getId()).thenReturn(UUID.randomUUID());
        when(role.isActive()).thenReturn(true);
        when(roleRepository.findByCode(roleCode)).thenReturn(role);

        when(apiClientRepository.save(any())).thenReturn(mock(ApiClient.class));

        CreateApiClientResponse response = useCase.execute(command);

        assertEquals(tenantId, response.tenantId());
        verify(apiClientRepository, times(1)).save(any());
        verify(apiClientRoleRepository, times(1)).save(any());
    }

    @Test
    void mustDenyPlatformAdminRoleEvenWhenNotFirstInTheList() {
        UUID tenantId = UUID.randomUUID();
        RoleCode commonCode = new RoleCode("customers.viewer");
        RoleCode adminCode = new RoleCode("platform.admin");
        CreateApiClientCommand command = new CreateApiClientCommand(
            tenantId, "ERP-client", null, List.of(commonCode, adminCode), List.of("identity.admin"));

        Tenant tenant = mock(Tenant.class);
        when(tenant.isActive()).thenReturn(true);
        when(tenantRepository.findById(tenantId)).thenReturn(tenant);

        Role commonRole = mock(Role.class);
        when(commonRole.getCode()).thenReturn(commonCode);
        when(commonRole.getModuleId()).thenReturn(null);
        when(commonRole.isActive()).thenReturn(true);
        when(roleRepository.findByCode(commonCode)).thenReturn(commonRole);

        Role adminRole = mock(Role.class);
        when(adminRole.getCode()).thenReturn(adminCode);
        when(adminRole.getModuleId()).thenReturn(null);
        when(adminRole.isActive()).thenReturn(true);
        when(roleRepository.findByCode(adminCode)).thenReturn(adminRole);

        AuthorizationDeniedException ex = assertThrows(AuthorizationDeniedException.class, () -> useCase.execute(command));
        assertEquals("denied", ex.getMessage());

        verify(apiClientRepository, times(0)).save(any());
        verify(apiClientRoleRepository, times(0)).save(any());
    }

    @Test
    void mustAllowNonPlatformAdminToCreateApiClientWithCommonRole() {
        UUID tenantId = UUID.randomUUID();
        RoleCode roleCode = new RoleCode("customers.viewer");
        CreateApiClientCommand command = new CreateApiClientCommand(
            tenantId, "ERP-client", null, List.of(roleCode), List.of("identity.admin"));

        Tenant tenant = mock(Tenant.class);
        when(tenant.isActive()).thenReturn(true);
        when(tenantRepository.findById(tenantId)).thenReturn(tenant);

        UUID moduleId = UUID.randomUUID();
        Role role = mock(Role.class);
        when(role.getCode()).thenReturn(roleCode);
        when(role.getModuleId()).thenReturn(moduleId);
        when(role.getId()).thenReturn(UUID.randomUUID());
        when(role.isActive()).thenReturn(true);
        when(roleRepository.findByCode(roleCode)).thenReturn(role);

        TenantModule tenantModule = mock(TenantModule.class);
        when(tenantModule.getModuleId()).thenReturn(moduleId);
        when(tenantModuleRepository.findAllByTenantIdAndEnabled(tenantId, true)).thenReturn(List.of(tenantModule));

        when(apiClientRepository.save(any())).thenReturn(mock(ApiClient.class));

        CreateApiClientResponse response = useCase.execute(command);

        assertEquals(tenantId, response.tenantId());
        verify(apiClientRepository, times(1)).save(any());
        verify(apiClientRoleRepository, times(1)).save(any());
    }

    @Test
    void mustThrowRoleInactiveExceptionWhenRoleIsDeactivated() {
        UUID tenantId = UUID.randomUUID();
        RoleCode roleCode = new RoleCode("internal-payroll.admin");
        CreateApiClientCommand command = new CreateApiClientCommand(
            tenantId, "ERP-client", null, List.of(roleCode), List.of("identity.admin"));

        Tenant tenant = mock(Tenant.class);
        when(tenant.isActive()).thenReturn(true);
        when(tenantRepository.findById(tenantId)).thenReturn(tenant);

        Role role = mock(Role.class);
        when(role.isActive()).thenReturn(false);
        when(roleRepository.findByCode(roleCode)).thenReturn(role);

        assertThrows(RoleInactiveException.class, () -> useCase.execute(command));

        verify(apiClientRepository, times(0)).save(any());
    }
}

File: services/identity-service/src/test/kotlin/br/com/deltaglobalbank/identity/features/apiClient/createApiClient/CreateApiClientTests.kt
package br.com.deltaglobalbank.identity.features.apiClient.createApiClient

import br.com.deltaglobalbank.identity.domain.apiClient.ApiClientRepository
import br.com.deltaglobalbank.identity.domain.module.ModuleRepository
import br.com.deltaglobalbank.identity.domain.module.TenantModule
import br.com.deltaglobalbank.identity.domain.module.TenantModuleRepository
import br.com.deltaglobalbank.identity.domain.role.Role
import br.com.deltaglobalbank.identity.domain.role.RoleCode
import br.com.deltaglobalbank.identity.domain.role.RoleRepository
import br.com.deltaglobalbank.identity.domain.tenant.Tenant
import br.com.deltaglobalbank.identity.domain.tenant.TenantRepository
import br.com.deltaglobalbank.identity.domain.user.DuplicateRoleException
import br.com.deltaglobalbank.identity.domain.user.ModuleNotEnabledForTenantException
import br.com.deltaglobalbank.identity.domain.user.RoleNotFoundException
import br.com.deltaglobalbank.identity.domain.user.TenantInactiveException
import br.com.deltaglobalbank.identity.domain.user.TenantNotFoundException
import br.com.deltaglobalbank.identity.features.apiClients.createApiClient.CreateApiClientCommand
import br.com.deltaglobalbank.identity.features.apiClients.createApiClient.CreateApiClientUseCase
import br.com.deltaglobalbank.identity.infrastructure.persistence.repositories.JpaApiClientRoleRepository

import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.mockk
import io.mockk.verify
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertAll
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.security.authorization.AuthorizationDeniedException
import java.util.UUID

@ExtendWith(MockKExtension::class)
class CreateApiClientTests {

    @MockK
    lateinit var apiClientRepository: ApiClientRepository

    @MockK
    lateinit var tenantRepository: TenantRepository

    @MockK
    lateinit var roleRepository: RoleRepository

    @MockK(relaxed = true)
    lateinit var tenantModuleRepository: TenantModuleRepository

    @MockK(relaxed = true)
    lateinit var moduleRepository: ModuleRepository

    @MockK
    lateinit var apiClientRoleRepository: JpaApiClientRoleRepository

    private lateinit var useCase: CreateApiClientUseCase

    @BeforeEach
    fun setUp() {
        useCase = CreateApiClientUseCase(
            apiClientRepository,
            tenantRepository,
            roleRepository,
            tenantModuleRepository,
            moduleRepository,
            apiClientRoleRepository
        )
    }

    @Test
    fun `must create api client successfully with roles`() {
        val tenantId = UUID.randomUUID()
        val roleCode = RoleCode("customers.admin")
        val command = CreateApiClientCommand(
            tenantId = tenantId,
            name = "ERP-client",
            description = "ERP",
            roleCodes = listOf(roleCode),
            creatorRoles = listOf("platform.admin")
        )

        val tenant = mockk<Tenant>()
        every { tenant.isActive() } returns true
        every { tenantRepository.findById(tenantId) } returns tenant

        val moduleId = UUID.randomUUID()
        val role = mockk<Role>()
        every { role.code } returns roleCode
        every { role.moduleId } returns moduleId
        every { role.id } returns UUID.randomUUID()
        every { roleRepository.findByCode(roleCode) } returns role

        val tenantModule = mockk<TenantModule>()
        every { tenantModule.moduleId } returns moduleId
        every { tenantModuleRepository.findAllByTenantIdAndEnabled(tenantId, true) } returns listOf(
            tenantModule
        )

        every { apiClientRepository.save(any()) } returns mockk(relaxed = true)
        every { apiClientRoleRepository.save(any()) } returns mockk(relaxed = true)

        val response = useCase.execute(command)

        assertAll(
            { assertEquals(tenantId, response.tenantId) },
            { assertEquals("ERP-client", response.name) },
            { assertEquals("ERP", response.description) }
        )

        verify(exactly = 1) { apiClientRepository.save(any()) }
        verify(exactly = 1) { apiClientRoleRepository.save(any()) }
    }

    @Test
    fun `must throw TenantNotFoundException when tenant does not exist`() {
        val command = CreateApiClientCommand(
            tenantId = UUID.randomUUID(),
            name = "ERP-client",
            description = null,
            roleCodes = listOf(RoleCode("customers.admin")),
            creatorRoles = listOf("platform.admin")
        )

        every { tenantRepository.findById(any()) } returns null

        assertThrows(TenantNotFoundException::class.java) {
            useCase.execute(command)
        }

        verify(exactly = 0) { apiClientRepository.save(any()) }
        verify(exactly = 0) { apiClientRoleRepository.save(any()) }
    }

    @Test
    fun `must throw TenantInactiveException when tenant is inactive`() {
        val tenantId = UUID.randomUUID()
        val command = CreateApiClientCommand(
            tenantId = tenantId,
            name = "ERP-client",
            description = null,
            roleCodes = listOf(RoleCode("customers.admin")),
            creatorRoles = listOf("platform.admin")
        )

        val tenant = mockk<Tenant>()
        every { tenant.isActive() } returns false
        every { tenantRepository.findById(tenantId) } returns tenant

        assertThrows(TenantInactiveException::class.java) {
            useCase.execute(command)
        }

        verify(exactly = 0) { apiClientRepository.save(any()) }
        verify(exactly = 0) { apiClientRoleRepository.save(any()) }
    }

    @Test
    fun `must throw DuplicateRoleException when role codes are exactly duplicated`() {
        val tenantId = UUID.randomUUID()
        val roleCode = RoleCode("customers.admin")
        val command = CreateApiClientCommand(
            tenantId = tenantId,
            name = "ERP-client",
            description = null,
            roleCodes = listOf(roleCode, roleCode),
            creatorRoles = listOf("platform.admin")
        )

        val tenant = mockk<Tenant>()
        every { tenant.isActive() } returns true
        every { tenantRepository.findById(tenantId) } returns tenant

        assertThrows(DuplicateRoleException::class.java) {
            useCase.execute(command)
        }

        verify(exactly = 0) { apiClientRepository.save(any()) }
        verify(exactly = 0) { apiClientRoleRepository.save(any()) }
    }

    @Test
    fun `must throw DuplicateRoleException when role codes differ only by case`() {
        val tenantId = UUID.randomUUID()
        val command = CreateApiClientCommand(
            tenantId = tenantId,
            name = "ERP-client",
            description = null,
            roleCodes = listOf(
                RoleCode("CUSTOMERS.ADMIN"),
                RoleCode("customers.admin")
            ),
            creatorRoles = listOf("platform.admin")
        )

        val tenant = mockk<Tenant>()
        every { tenant.isActive() } returns true
        every { tenantRepository.findById(tenantId) } returns tenant

        assertThrows(DuplicateRoleException::class.java) {
            useCase.execute(command)
        }

        verify(exactly = 0) { apiClientRepository.save(any()) }
        verify(exactly = 0) { apiClientRoleRepository.save(any()) }
    }

    @Test
    fun `must throw RoleNotFoundException when role does not exist`() {
        val tenantId = UUID.randomUUID()
        val roleCode = RoleCode("nonexistent.role")
        val command = CreateApiClientCommand(
            tenantId = tenantId,
            name = "ERP-client",
            description = null,
            roleCodes = listOf(roleCode),
            creatorRoles = listOf("platform.admin")
        )

        val tenant = mockk<Tenant>()
        every { tenant.isActive() } returns true
        every { tenantRepository.findById(tenantId) } returns tenant

        every { roleRepository.findByCode(roleCode) } returns null

        assertThrows(RoleNotFoundException::class.java) {
            useCase.execute(command)
        }

        verify(exactly = 0) { apiClientRepository.save(any()) }
        verify(exactly = 0) { apiClientRoleRepository.save(any()) }
    }

    @Test
    fun `must throw ModuleNotEnabledForTenantException when role module is not enabled`() {
        val tenantId = UUID.randomUUID()
        val roleCode = RoleCode("customers.admin")
        val command = CreateApiClientCommand(
            tenantId = tenantId,
            name = "ERP-client",
            description = null,
            roleCodes = listOf(roleCode),
            creatorRoles = listOf("platform.admin")
        )

        val tenant = mockk<Tenant>()
        every { tenant.isActive() } returns true
        every { tenantRepository.findById(tenantId) } returns tenant

        val moduleId = UUID.randomUUID()
        val role = mockk<Role>()
        every { role.code } returns roleCode
        every { role.moduleId } returns moduleId
        every { roleRepository.findByCode(roleCode) } returns role

        every {
            tenantModuleRepository.findAllByTenantIdAndEnabled(
                tenantId,
                true
            )
        } returns emptyList()
        every { moduleRepository.findById(moduleId) } returns null

        assertThrows(ModuleNotEnabledForTenantException::class.java) {
            useCase.execute(command)
        }

        verify(exactly = 0) { apiClientRepository.save(any()) }
        verify(exactly = 0) { apiClientRoleRepository.save(any()) }
    }

    @Test
    fun `must deny when non-platform-admin creates api client with platform-admin role`() {
        val tenantId = UUID.randomUUID()
        val roleCode = RoleCode("platform.admin")
        val command = CreateApiClientCommand(
            tenantId = tenantId,
            name = "ERP-client",
            description = null,
            roleCodes = listOf(roleCode),
            creatorRoles = listOf("identity.admin")
        )

        val tenant = mockk<Tenant>()
        every { tenant.isActive() } returns true
        every { tenantRepository.findById(tenantId) } returns tenant

        val role = mockk<Role>()
        every { role.code } returns roleCode
        every { role.moduleId } returns null
        every { roleRepository.findByCode(roleCode) } returns role

        val ex = assertThrows(AuthorizationDeniedException::class.java) {
            useCase.execute(command)
        }
        assertEquals("denied", ex.message)

        verify(exactly = 0) { apiClientRepository.save(any()) }
        verify(exactly = 0) { apiClientRoleRepository.save(any()) }
    }

    @Test
    fun `must allow platform-admin to create api client with platform-admin role`() {
        val tenantId = UUID.randomUUID()
        val roleCode = RoleCode("platform.admin")
        val command = CreateApiClientCommand(
            tenantId = tenantId,
            name = "ERP-client",
            description = null,
            roleCodes = listOf(roleCode),
            creatorRoles = listOf("platform.admin")
        )

        val tenant = mockk<Tenant>()
        every { tenant.isActive() } returns true
        every { tenantRepository.findById(tenantId) } returns tenant

        val role = mockk<Role>()
        every { role.code } returns roleCode
        every { role.moduleId } returns null
        every { role.id } returns UUID.randomUUID()
        every { roleRepository.findByCode(roleCode) } returns role

        every { apiClientRepository.save(any()) } returns mockk(relaxed = true)
        every { apiClientRoleRepository.save(any()) } returns mockk(relaxed = true)

        val response = useCase.execute(command)

        assertEquals(tenantId, response.tenantId)
        verify(exactly = 1) { apiClientRepository.save(any()) }
        verify(exactly = 1) { apiClientRoleRepository.save(any()) }
    }

    @Test
    fun `must deny platform-admin role even when not first in the list`() {
        val tenantId = UUID.randomUUID()
        val commonCode = RoleCode("customers.viewer")
        val adminCode = RoleCode("platform.admin")
        val command = CreateApiClientCommand(
            tenantId = tenantId,
            name = "ERP-client",
            description = null,
            roleCodes = listOf(commonCode, adminCode),
            creatorRoles = listOf("identity.admin")
        )

        val tenant = mockk<Tenant>()
        every { tenant.isActive() } returns true
        every { tenantRepository.findById(tenantId) } returns tenant

        val commonRole = mockk<Role>()
        every { commonRole.code } returns commonCode
        every { commonRole.moduleId } returns null
        every { roleRepository.findByCode(commonCode) } returns commonRole

        val adminRole = mockk<Role>()
        every { adminRole.code } returns adminCode
        every { adminRole.moduleId } returns null
        every { roleRepository.findByCode(adminCode) } returns adminRole

        val ex = assertThrows(AuthorizationDeniedException::class.java) {
            useCase.execute(command)
        }
        assertEquals("denied", ex.message)

        verify(exactly = 0) { apiClientRepository.save(any()) }
        verify(exactly = 0) { apiClientRoleRepository.save(any()) }
    }

    @Test
    fun `must allow non-platform-admin to create api client with common role`() {
        val tenantId = UUID.randomUUID()
        val roleCode = RoleCode("customers.viewer")
        val command = CreateApiClientCommand(
            tenantId = tenantId,
            name = "ERP-client",
            description = null,
            roleCodes = listOf(roleCode),
            creatorRoles = listOf("identity.admin")
        )

        val tenant = mockk<Tenant>()
        every { tenant.isActive() } returns true
        every { tenantRepository.findById(tenantId) } returns tenant

        val moduleId = UUID.randomUUID()
        val role = mockk<Role>()
        every { role.code } returns roleCode
        every { role.moduleId } returns moduleId
        every { role.id } returns UUID.randomUUID()
        every { roleRepository.findByCode(roleCode) } returns role

        val tenantModule = mockk<TenantModule>()
        every { tenantModule.moduleId } returns moduleId
        every { tenantModuleRepository.findAllByTenantIdAndEnabled(tenantId, true) } returns listOf(
            tenantModule
        )

        every { apiClientRepository.save(any()) } returns mockk(relaxed = true)
        every { apiClientRoleRepository.save(any()) } returns mockk(relaxed = true)

        val response = useCase.execute(command)

        assertEquals(tenantId, response.tenantId)
        verify(exactly = 1) { apiClientRepository.save(any()) }
        verify(exactly = 1) { apiClientRoleRepository.save(any()) }
    }

}